### PR TITLE
[Brave News]: Change top news card to show top topics

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/35319

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_news/browser/feed_v2_builder.cc [hunk 1]
```diff
@@ -1235,8 +1271,9 @@ mojom::FeedV2Ptr FeedV2Builder::GenerateAllFeed() {
 
   // Step 3: Generate a top news block
   // https://docs.google.com/document/d/1bSVHunwmcHwyQTpa3ab4KRbGbgNQ3ym_GHvONnrBypg/edit#heading=h.7z05nb4b269d
-  auto top_news_block =
-      GenerateChannelBlock(locale, publishers, kTopNewsChannel, articles);
+  // This block is a bit special - we take the top articles from the top few
+  // topics and display them in a cluster.
+  auto top_news_block = GenerateTopTopicsBlock(publishers, topics);
```
Comment (LorenzoMinto) [inline components/brave_news/browser/feed_v2_builder.cc:1276]: this will only happen on the third block of the For You feed, right?

Comment (LorenzoMinto) [reply components/brave_news/browser/feed_v2_builder.cc:1276]: and only once

Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_builder.cc:1276]: Yup, there'll be 1 Normal Block, one Ad, then this.

Comment (LorenzoMinto) [reply components/brave_news/browser/feed_v2_builder.cc:1276]: what happens if topics are empty? Like in most locales

Comment (LorenzoMinto) [reply components/brave_news/browser/feed_v2_builder.cc:1276]: Oh I see, we return an empty block. Could fall back to top news channel cluster if topics are empty instead?

**Review by LorenzoMinto — APPROVED** [review]
Just need to handle the empty topics scenario better (non-US locales don't have topics right now). Otherwise lgtm. Thanks @fallaciousreasoning 🙏

File: components/brave_news/browser/feed_v2_builder.cc [hunk 2]
```diff
@@ -599,6 +599,55 @@ std::vector<mojom::FeedItemV2Ptr> GenerateChannelBlock(
...
+    }
+
+    auto item = FromTopicArticle(publishers, articles.at(0));
+    items.push_back(mojom::ArticleElements::NewArticle(
+        mojom::Article::New(std::move(item), false)));
+    if (items.size() >
+        static_cast<size_t>(features::kBraveNewsMaxBlockCards.Get())) {
+      break;
+    }
+  }
```
Comment (simonhong) [inline components/brave_news/browser/feed_v2_builder.cc:644]: nit: don't need to call features::kBraveNewsMaxBlockCards.Get() multiple times.
```suggestion
  std::vector<mojom::ArticleElementsPtr> items;
  const auto max_block_size =
      static_cast<size_t>(features::kBraveNewsMaxBlockCards.Get());
  for (const auto& [topic, articles] : topics) {
    if (articles.empty()) {
      continue;
    }

    auto item = FromTopicArticle(publishers, articles.at(0));
    items.push_back(mojom::ArticleElements::NewArticle(
        mojom::Article::New(std::move(item), false)));
    if (items.size() > max_block_size) {
      break;
    }
  }
```

**Review by simonhong — APPROVED** [review]
++

**Comment by fallaciousreasoning** [comment]
We fall back to the old behavior when we fail to generate the block from topics now :smile:

**Comment by stephendonner** [comment]
## Verified `IN-PROGRESS` using

```
Brave | 1.64.32 Chromium: 121.0.6167.85 (Official Build) nightly (64-bit)
-- | --
Revision | 59bd6e14828f1b3b3638729ff241d7c2055b1e38
OS | Windows 10 Version 22H2 (Build 19045.3996)
```

**Pull request merged**